File: tools/dotnet/Common/IO/Log.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DTY.Dotnet.Tools.Common.IO
{
    public enum LogLevel
    {
        INFO,
        WARN,
        ERROR,
        FATAL
    }

    public interface ILog
    {
        public bool IsDebugMode { get; set; }
        public bool IsConsoleOutput { get; set; }
        public string Id { get; }
        public string Name { get; }
        public string Description { get; }
        public string Path { get; }

        public void StartLog(bool append);
        public void EndLog();
        public void Logging(LogLevel level, string message);
        public void Logging(LogLevel level, DateTime time, string message);
        public void Logging(string msg = "");
    }

    internal class Log : ILog
    {
        public bool IsDebugMode { get; set; }
        public bool IsConsoleOutput { get; set; }

        public string Id { get; }
        public string Name { get; }
        public string Description { get; }
        public string Path { get; }

        private StreamWriter? _Writer;

        private Log(string name, FileInfo path, string des = "")
        {
            IsConsoleOutput = false;

            Id = Guid.NewGuid().ToString();
            Name = name;
            Path = path.FullName;
            Description = des;

            _Writer = null;
        }

        public void StartLog(bool append)
        {
            EndLog();

            bool fileExist = File.Exists(Path);

            _Writer = new StreamWriter(Path, append, Encoding.UTF8);

            if (!fileExist || !append)
            {
                _Writer?.WriteLine(Name);
                _Writer?.WriteLine(string.Format("说明：{0}", Description));
                _Writer?.WriteLine(string.Format("ID：{0}", Id));
                _Writer?.WriteLine();

                _Writer?.Flush();
            }
        }

        public void EndLog()
        {
            _Writer?.Close();
            _Writer?.Dispose();
            _Writer = null;
        }

        public void Logging(LogLevel level, string message)
        {
            Logging(level, DateTime.Now, message);
        }

        public void Logging(LogLevel level, DateTime time, string message)
        {
            string levelStr = level switch
            {
                LogLevel.INFO => "INFO ",
                LogLevel.WARN => "WARN ",
                LogLevel.ERROR => "ERROR",
                LogLevel.FATAL => "FATAL",
                _ => "INFO "
            };
            string timeStr = time.ToString("yyyy-MM-dd HH:mm:ss.fff");

            string formated_msg = string.Format("[{0}] [{1}] {2}", levelStr, timeStr, message);
            Logging_Internal(level, formated_msg);
        }

        public void Logging(string msg = "")
        {
            Logging_Internal(LogLevel.INFO, msg);

            if (IsConsoleOutput)
                Console.WriteLine(msg);
        }

        private void Logging_Internal(LogLevel level, string msg)
        {
            if (!IsDebugMode && LogLevel.ERROR != level && LogLevel.FATAL != level)
                return;

            _Writer?.WriteLine(msg);
            _Writer?.Flush();

            if (IsConsoleOutput)
                WriteToConsole(level, msg);
        }

        private void WriteToConsole(LogLevel color, string formated_msg)
        {
            Console.ForegroundColor = color switch
            {
                LogLevel.ERROR => ConsoleColor.Red,
                LogLevel.INFO => ConsoleColor.Green,
                LogLevel.WARN => ConsoleColor.Yellow,
                LogLevel.FATAL => ConsoleColor.Magenta,
                _ => ConsoleColor.Green,
            };

            Console.WriteLine(formated_msg);

            Console.ForegroundColor = ConsoleColor.White;
        }

        public static Log GenerateLogger(string name, string path, string des = "")
        {
            if (name is null || path is null)
                throw new ArgumentNullException(nameof(GenerateLogger));

            if (name.Trim().Length == 0)
                throw new ArgumentException(nameof(name));

            FileInfo info = new(path);

            return new Log(name, info, des);
        }
    }
}

File: .project/derelict/tools/java/Tools/Common/IO/Log.java
package Tools.Common.IO;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.UUID;

import Tools.Common.LogLevel;

public class Log implements ILog
{
    private static final DateFormat LogDateFormatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
    private static final String ERRORConsoleFormat = "\033[31;4m%s\033[0m";
    private static final String WARNConsoleFormat = "\033[33;4m%s\033[0m";
    private static final String INFOConsoleFormat = "\033[32;4m%s\033[0m";
    private static final String FATALConsoleFormat = "\033[46;37;4m%s\033[0m";

    private final UUID _Id;
    private final String _Name;
    private final String _Path;
    private final String _Description;

    private boolean _IsDebugMode;
    private boolean _IsConsoleOutput;
    private FileWriter _Writer;

    private Log(String name, String path, String des)
    {
        _Id = UUID.randomUUID();
        _Name = name;
        _Path = path;
        _Description = des;

        _IsDebugMode = false;
        _IsConsoleOutput = false;
        _Writer = null;
    }

    @Override
    public String GetId()
    {
        return this._Id.toString();
    }

    @Override
    public String GetName()
    {
        return this._Name;
    }

    @Override
    public String GetDescription()
    {
        return this._Description;
    }

    @Override
    public String GetPath()
    {
        return this._Path;
    }

    @Override
    public boolean GetIsDebugMode()
    {
        return this._IsDebugMode;
    }

    @Override
    public boolean GetIsConsoleOutput()
    {
        return this._IsConsoleOutput;
    }

    @Override
    public void SetIsDebugMode(boolean debugMode)
    {
        this._IsDebugMode = debugMode;
    }

    @Override
    public void SetIsConsoleOutput(boolean isConsoleOutput)
    {
        this._IsConsoleOutput = isConsoleOutput;
    }

    @Override
    public void StartLog(boolean append) throws IOException
    {
        this.EndLog();

        File file = new File(this._Path);
        boolean fileExist = file.exists();

        this._Writer = new FileWriter(file, append);

        if (!fileExist || !append)
        {
            this._Writer.write(String.format("%s%s", this._Name, System.getProperty("line.separator")));
            this._Writer.write(String.format("说明：%s%s", this._Description, System.getProperty("line.separator")));
            this._Writer.write(String.format("ID：%s%s", this._Id, System.getProperty("line.separator")));
            this._Writer.write(System.getProperty("line.separator"));

            this._Writer.flush();
        }
    }

    @Override
    public void EndLog() throws IOException
    {
        if (null == this._Writer)
            return;

        this._Writer.close();
        this._Writer = null;
    }

    @Override
    public void Logging(LogLevel level, String message) throws IOException
    {
        this.Logging(level, new Date(), message);
    }

    @Override
    public void Logging(LogLevel level, Date time, String message) throws IOException
    {
        String levelStr = "";

        switch (level)
        {
            case INFO:
                levelStr = "INFO ";
                break;
            case WARN:
                levelStr = "WARN ";
                break;
            case ERROR:
                levelStr = "ERROR";
                break;
            case FATAL:
                levelStr = "FATAL";
                break;
            default:
                levelStr = "INFO ";
                break;
        }

        String timeStr = Log.LogDateFormatter.format(time);
        String formated_msg = String.format("[%s] [%s] %s", levelStr, timeStr, message);
        this.Logging_Internal(level, formated_msg);
    }

    @Override
    public void Logging(String msg) throws IOException
    {
        this.Logging_Internal(LogLevel.INFO, msg);
    }

    private void Logging_Internal(LogLevel level, String msg) throws IOException
    {
        if (null == this._Writer)
            return;

        if (!this._IsDebugMode && LogLevel.ERROR != level && LogLevel.FATAL != level)
            return;

        this._Writer.write(msg);
        this._Writer.write(System.getProperty("line.separator"));
        this._Writer.flush();

        if (this._IsConsoleOutput)
            WriteToConsole(level, msg);
    }

    private void WriteToConsole(LogLevel color, String formated_msg)
    {
        String formatStr = "";

        switch (color)
        {
            case WARN:
                formatStr = Log.WARNConsoleFormat;
                break;
            case ERROR:
                formatStr = Log.ERRORConsoleFormat;
                break;
            case FATAL:
                formatStr = Log.FATALConsoleFormat;
                break;
            case INFO:
            default:
                formatStr = Log.INFOConsoleFormat;
                break;
        }

        System.out.println(String.format(formatStr, formated_msg));
    }

    public static Log GenerateLogger(String name, String path) throws Exception
    {
        return Log.GenerateLogger(name, path, "");
    }

    public static Log GenerateLogger(String name, String path, String des) throws Exception
    {
        if (null == name || null == path)
            throw new Exception("GenerateLogger");

        if (name.isEmpty() || name.isBlank())
            throw new Exception("name");

        return new Log(name, path, des);
    }
}